chore: Upgrade Next.js from v12 to v14

tihuan opened the pull request:
## Reason for Change

- #TICKET_NUMBER
- If the reason for this PR's code changes are not clear in the issue, state value/impact

## Changes

- add
- remove
- modify

## Testing steps

- Either list QA steps or reasoning you feel QA is unnecessary
- Describe how you made sure to know that your changes worked. Should allow someone else to go verify your code without in depth knowledge.
- "Unit tested only", "tested in rdev by a, b, c, verifying feature worked by... ", "manually ran pipeline locally with these results: ..."

## Checklist 🛎️

- [ ] Add product, design, and eng as reviewers for rdev review
- [ ] For UI changes, add screenshots/videos, so the reviewers know what you expect them to see
- [ ] For UI changes, add e2e tests to prevent regressions

## Notes for Reviewer

tihuan commented on frontend/src/views/WheresMyGeneV2/components/HeatMap/connect.ts:
```diff
@@ -211,6 +211,10 @@ export function useConnect({
+       * WARNING: Doesn't pass React StrictMode check
+       * We get double events in dev mode, will need further investigation
+       */
       dispatch?.(toggleExpandedTissueId({ tissueId, tissueName }));
```
something we need to keep an eye on, since this code is not compatible with React StrictMode and future versions of React

tihuan commented on frontend/tests/utils/wmgUtils.ts:
```diff
@@ -257,38 +261,38 @@ export const selectTissueAndGeneOption = async (page: Page) => {
 export const checkSourceData = async (page: Page) => {
-  //click on source data icon
-  await page.locator('[data-testid="source-data-button"]').click();
+  const sourceDataButton = page.getByTestId(SOURCE_DATA_BUTTON_ID);
+  const sourceDataList = page.getByTestId(SOURCE_DATA_LIST_ID);
```
use the constants as single source of truth

tihuan commented on frontend/tests/utils/wmgUtils.ts:
```diff
@@ -257,38 +261,38 @@ export const selectTissueAndGeneOption = async (page: Page) => {
    */
   await tryUntil(
     async () => {
-      await page.keyboard.press("Escape");
-
```
update test steps here, since the new version of Blueprint doesn't work with Escape reliably

tihuan commented on frontend/tsconfig.json:
```diff
@@ -5,24 +5,48 @@
     "forceConsistentCasingInFileNames": true,
     "resolveJsonModule": true,
-    "incremental": true
+    "incremental": true,
+    "plugins": [
+      {
+        "name": "next"
+      }
+    ]
```
actual change from Next.js upgrade

tihuan commented on frontend/tsconfig.json:
```diff
@@ -5,24 +5,48 @@
+    "**/*.ts",
+    "**/*.tsx",
+    "**/*.js",
+    ".next/types/**/*.ts"
```
actual change from Next.js upgrade

tihuan commented on frontend/tsconfig.json:
```diff
@@ -5,24 +5,48 @@
     "esModuleInterop": true,
```
lots of prettier changes

frano-m commented on frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/index.tsx:
```diff
@@ -236,7 +236,6 @@ const DatasetRow: FC<Props> = ({
           />
           {hasCXGFile(dataset) && (
             <Tooltip
-              boundary="viewport"
```
Hi @tihuan, this looks good to me! As a side note, I am seeing a warning in the console - when hovering over the NTag. Are you seeing this as well?

`client.js:1 Warning: Legacy context API has been detected within a strict-mode tree. The old API will be supported in all 16.x releases, but applications using it should migrate to the new version. Please update the following components: Blueprint5.Portal`

frano-m commented on frontend/src/components/CreateCollectionModal/index.tsx:
```diff
@@ -24,7 +25,7 @@ const AsyncCTA = loadable(
 );
 
-const CreateCollectionButton = (props: Partial<IButtonProps>) => (
+const CreateCollectionButton = (props: Partial<ButtonProps>) => (
```
Nice pick up thanks @tihuan! ⭐

frano-m commented on frontend/src/components/common/MoreDropdown/index.tsx:
```diff
@@ -1,14 +1,15 @@
-  popoverProps?: IPopoverProps;
-  buttonProps?: IButtonProps;
+  popoverProps?: PopoverProps;
+  buttonProps?: Partial<ButtonIconProps<"dotsHorizontal", "small">>;
```
@tihuan thank you 🎆

frano-m commented:
Hi @tihuan, I'm seeing a warning in my terminal. Are you seeing this one too?

`Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.`

tihuan commented:
Thanks so much for the thorough testing, @frano-m ! Sorry that I missed those warnings 🙏 I'll fix them today!

I can repro the NTag warning too 🙆‍♂️ 

<img width="1425" alt="Screenshot 2024-01-18 at 7 51 32 AM" src="https://github.com/chanzuckerberg/single-cell-data-portal/assets/6309723/754c6c98-430f-4a2c-8711-ffafeeaa27de">

<img width="1009" alt="Screenshot 2024-01-18 at 7 51 45 AM" src="https://github.com/chanzuckerberg/single-cell-data-portal/assets/6309723/2f00dc14-b65c-422b-876d-f1ed07e35461">

tihuan commented:
@frano-m I fixed the layout effect error (caused by using Blueprint AnchorButton`) and replaced `:first-child` with `:first-of-type`, and the errors seemed to go away now!

The `Blueprint5.Portal` error I think it's because Blueprintv5 is still supporting React Legacy Context API and only marking it as deprecated. So I think the error naturally go away once Blueprint removes the support in v6!

tihuan commented on frontend/src/components/Header/components/Nav/index.tsx:
```diff
@@ -24,58 +23,42 @@ export default function Nav({ className, pathname }: Props): JSX.Element {
           <NavItemContainer>
-            <LinkWrapper>
-              <Link href={ROUTES.COLLECTIONS} passHref>
-                <AnchorButton
```
@seve I removed `AnchorButton` usage since it violates React 18 usage. I think the nav link still works after I updated the style. But PTAL in case I missed anything 🙏 Thank you!

tihuan commented:
Thanks so much for the review, @seve ! Will wait for @frano-m to approve before merging 👍 !

frano-m reviewed (approved):
LGTM @tihuan! Nice work and thank you 🚀

tihuan commented:
Amazing! Thanks so much again for catching the errors, @frano-m 🤩 🙏 !
